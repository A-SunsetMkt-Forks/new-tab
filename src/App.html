<script>
  import { handleLinkClick } from './common.js';
  import BookmarkGroup from './components/BookmarkGroup.html';
  import Menu from './components/Menu.html';
  import Search from './components/Search.html';

  window.addEventListener('click', handleLinkClick, true);
</script>

<BookmarkGroup />
<Search />
<Menu />

<style type="text/postcss">
  @import './themes/light.css';
  @import './themes/dark.css';

  :global(body) {
    margin: 80px 0;
    /* overflow-y: scroll; */ /* prevents jump on load */
    font-size: 18px;
    /**
     * Normally you would use a unitless value for line-height but this is kind
     * of a hack for alignment with the minimum possible style properties. Same
     * as bookmark bar height.
     */
    line-height: 41px;
    text-rendering: optimizeSpeed;
    font-variant-ligatures: none;
    color: var(--t);
    background-color: var(--b);
  }

  :global(h2),
  :global(.muted) {
    color: var(--c1);
  }

  :global(a) { /* stylelint-disable-line no-descending-specificity */
    text-decoration: none;

    &:hover,
    &:focus {
      text-decoration: underline;
    }
  }
</style>
