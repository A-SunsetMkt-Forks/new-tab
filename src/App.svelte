<script>
  import { handleLinkClick } from './utils';
  import BookmarksBar from './components/BookmarksBar.svelte';
  import Menu from './components/Menu.svelte';
  import Search from './components/Search.svelte';
</script>

<style type="text/postcss">
  @import './themes/light.css';
  @import './themes/dark.css';

  :global(body) { /* stylelint-disable-line a11y/line-height-is-vertical-rhythmed */
    margin: 110px 0;
    overflow-y: scroll; /* Prevent jump on load */
    font-size: 18px;
    /**
     * Normally you would use a unitless value for line-height but this is kind
     * of a hack for alignment with the minimum possible style properties. Same
     * as bookmark bar height.
     */
    line-height: 41px;
    text-rendering: optimizeSpeed;
    font-variant-ligatures: none;
    background-color: var(--b);
  }

  :global(h2),
  :global(.muted) {
    color: var(--c1);
  }

  :global(a) { /* stylelint-disable-line no-descending-specificity */
    text-decoration: none;

    &:hover,
    &:focus {
      text-decoration: underline;
    }
  }
</style>

<svelte:window on:click={handleLinkClick}/>

<BookmarksBar/>
<Search/>
<Menu/>
