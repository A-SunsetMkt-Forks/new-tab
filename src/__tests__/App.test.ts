// FIXME: Move things to `common.test.js` + fix click handler tests

// FIXME: Should tests be split up into virtual (jsdom) + UI (puppeteer)?

import App from '../App.svelte';

beforeAll(async () => {
  await page.goto('chrome://newtab');
});

describe('App component', () => {
  it('renders correctly', () => {
    const target = document.createElement('div');
    new App({ target });
    expect(target.innerHTML).toMatchSnapshot();
  });
});

describe('App click handler', () => {
  it.skip('triggers on mouse click event', () => {
    expect.assertions(1);
    const target = document.createElement('div');
    const spy1 = jest.spyOn(chrome.tabs, 'update');
    new App({ target });
    const event = new MouseEvent('click');
    Object.defineProperty(event, 'target', {
      enumerable: true,
      value: { href: 'chrome://bookmarks/' },
    });
    const spy2 = jest.spyOn(event, 'preventDefault');
    window.dispatchEvent(event);
    expect(spy1).toHaveBeenCalledWith(event);
    expect(spy1).toHaveBeenCalledTimes(1);
    expect(spy2).toHaveBeenCalledTimes(1);
    spy1.mockRestore();
    spy2.mockRestore();
  });

  it.skip('handles regular HTTP links', () => {
    expect.assertions(2);
    const target = document.createElement('div');
    const component = new App({ target });
    const link = target.querySelector('[href="https://github.com/MaxMilton/new-tab/issues"]');
    const event = new MouseEvent('click', {
      bubbles: true,
      cancelable: true,
      view: window,
    });
    Object.defineProperty(event, 'target', { enumerable: true, value: link });
    const spy1 = jest.spyOn(component, '_onLinkClick');
    const spy2 = jest.spyOn(event, 'preventDefault');
    window.dispatchEvent(event);
    expect(spy1).toHaveBeenCalled();
    expect(spy2).not.toHaveBeenCalled();
    spy1.mockRestore();
    spy2.mockRestore();
  });

  it.skip('handles chrome internal links', () => {
    expect.assertions(4);
    const target = document.createElement('div');
    const component = new App({ target });
    const link = target.querySelector('[href="chrome://bookmarks/"]');
    const event = new MouseEvent('click', {
      bubbles: true,
      cancelable: true,
      view: window,
    });
    Object.defineProperty(event, 'target', { enumerable: true, value: link });
    const spy1 = jest.spyOn(component, '_onLinkClick');
    const spy2 = jest.spyOn(event, 'preventDefault');
    const spy3 = jest.spyOn(chrome.tabs, 'create', 'get');
    const spy4 = jest.spyOn(chrome.tabs, 'update', 'get');
    window.dispatchEvent(event);
    expect(spy1).toHaveBeenCalled();
    expect(spy2).toHaveBeenCalled();
    expect(spy3).not.toHaveBeenCalled();
    expect(spy4).toHaveBeenCalled();
    spy1.mockRestore();
    spy2.mockRestore();
    spy3.mockRestore();
    spy4.mockRestore();
  });

  it.skip('handles chrome internal links with _blank target', () => {
    expect.assertions(4);
    const target = document.createElement('div');
    const component = new App({ target });
    const link = target.querySelector('[href="chrome://bookmarks/"]');
    const event = new MouseEvent('click', {
      bubbles: true,
      cancelable: true,
      view: window,
    });
    Object.defineProperty(event, 'target', { enumerable: true, value: link });
    Object.defineProperty(event.target, 'target', {
      enumerable: true,
      value: '_blank',
    });
    const spy1 = jest.spyOn(component, '_onLinkClick');
    const spy2 = jest.spyOn(event, 'preventDefault');
    const spy3 = jest.spyOn(chrome.tabs, 'create', 'get');
    const spy4 = jest.spyOn(chrome.tabs, 'update', 'get');
    window.dispatchEvent(event);
    expect(spy1).toHaveBeenCalled();
    expect(spy2).toHaveBeenCalled();
    expect(spy3).toHaveBeenCalled();
    expect(spy4).not.toHaveBeenCalled();
    spy1.mockRestore();
    spy2.mockRestore();
    spy3.mockRestore();
    spy4.mockRestore();
  });

  it.skip('handles chrome internal links with ctrl key', () => {
    expect.assertions(4);
    const target = document.createElement('div');
    const component = new App({ target });
    const link = target.querySelector('[href="chrome://bookmarks/"]');
    const event = new MouseEvent('click', {
      bubbles: true,
      cancelable: true,
      ctrlKey: true,
      view: window,
    });
    Object.defineProperty(event, 'target', { enumerable: true, value: link });
    const spy1 = jest.spyOn(component, '_onLinkClick');
    const spy2 = jest.spyOn(event, 'preventDefault');
    const spy3 = jest.spyOn(chrome.tabs, 'create', 'get');
    const spy4 = jest.spyOn(chrome.tabs, 'update', 'get');
    window.dispatchEvent(event);
    expect(spy1).toHaveBeenCalled();
    expect(spy2).toHaveBeenCalled();
    expect(spy3).toHaveBeenCalled();
    expect(spy4).not.toHaveBeenCalled();
    spy1.mockRestore();
    spy2.mockRestore();
    spy3.mockRestore();
    spy4.mockRestore();
  });
});
