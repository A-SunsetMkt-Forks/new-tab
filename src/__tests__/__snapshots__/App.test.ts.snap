// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component renders correctly 1`] = `
"<div class=\\"bookmarks\\">

  

  </div>
<div class=\\"container\\"><input id=\\"search\\" placeholder=\\"Search browser…\\" autocomplete=\\"off\\">

  </div>
<div id=\\"menu\\"><div id=\\"icon\\">☰</div>

  <div id=\\"dropdown\\"><a href=\\"chrome-search://local-ntp/local-ntp.html\\" class=\\"menu-item\\">Open Default Tab</a>
    <a href=\\"chrome://bookmarks/\\" class=\\"menu-item\\">Bookmarks Manager</a>
    <a href=\\"chrome://downloads/\\" class=\\"menu-item\\">Downloads</a>
    <a href=\\"chrome://history/\\" class=\\"menu-item\\">History</a>
    <a href=\\"chrome://settings/passwords\\" class=\\"menu-item\\">Passwords</a>

    <hr>

    
    <a id=\\"o\\" class=\\"menu-item\\">New Tab Settings</a>
    <a href=\\"https://github.com/MaxMilton/new-tab/issues\\" class=\\"menu-item\\">Submit Bug</a></div></div>"
`;
